`ptr::copy_nonoverlapping` is not memcpy

workingjubilee opened the pull request:
A C implementation may treat memcpy calls Rust expects `ptr::copy_nonoverlapping` to handle correctly as UB. Stop confusing the matter.

r? @RalfJung

workingjubilee committed:
`ptr::copy_nonoverlapping` is not memcpy

A C implementation may treat memcpy calls Rust expects
`ptr::copy_nonoverlapping` to handle correctly as UB.
Stop confusing the matter.

workingjubilee commented on library/core/src/intrinsics.rs:
```diff
@@ -2604,6 +2599,10 @@ pub(crate) fn is_nonoverlapping<T>(src: *const T, dst: *const T, count: usize) -
 /// Note that even if the effectively copied size (`count * size_of::<T>()`) is
 /// `0`, the pointers must be non-null and properly aligned.
 ///
+/// In the case of a 0-byte read, the pointers may be the same pointer,
+/// as a region of 0 bytes overlaps with a region of 0 bytes for 0 bytes.
```
i.e. it does not.

workingjubilee commented:
Also of interest:
- https://github.com/rust-lang/unsafe-code-guidelines/issues/426

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -2604,6 +2599,10 @@ pub(crate) fn is_nonoverlapping<T>(src: *const T, dst: *const T, count: usize) -
 ///
+/// In the case of a 0-byte read, the pointers may be the same pointer,
+/// as a region of 0 bytes overlaps with a region of 0 bytes for 0 bytes.
+/// A copy of 1 or more bytes to the same address incurs undefined behavior.
```
Yeah I wouldn't want to allow exactly overlapping copies for now. However, I am not sure if this comment is that helpful? We already have a paragraph above talking about the zero-size case; I'd just add the comment about overlaps there.

workingjubilee replied:
Like https://github.com/rust-lang/rust/pull/113347/commits/cb42825a322ffb5e34f658f25753891f2a26da81?

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -2574,14 +2574,9 @@ pub(crate) fn is_nonoverlapping<T>(src: *const T, dst: *const T, count: usize) -
 /// For regions of memory which might overlap, use [`copy`] instead.
 ///
-/// `copy_nonoverlapping` is semantically equivalent to C's [`memcpy`], but
-/// with the argument order swapped.
```
Is it worth saying that this is still the same as `memcpy` for non-zero-sized copies, or that it is similar to `memcpy`, or something like that?

RalfJung replied:
Many other programmers come from C though and it is helpful for them. :shrug: In particular since it is almost-but-not-quite the same.

workingjubilee replied:
The problem is that we have no policy here, so discussing the deviations in detail invites "why not talk about this similar function from `{lang}` until we have 10 different languages being discussed in what is supposed to be the documentation for a single function. T-libs has previously resolved to avoid this direction until a draft policy is proposed.

workingjubilee replied:
Also in actual cases, colleagues have noticed things like `write_bytes` saying "it's like memset" has caused programmers to **stop reading there**, and not even *think* about the disclaimer "...but it writes for this larger space", causing people who intended to "memset" a single page to instead thrash 16 pages because they didn't think about pointer type.

RalfJung commented:
Similar comments appear in `copy` and `write_bytes` (saying they are like `memmove` and `memset`), so those should also be updated.

workingjubilee commented:
Oh, *lovely*.

workingjubilee committed:
Nestle paragraphs in copy_nonoverlapping

workingjubilee committed:
Erase redundancy about 1+ bytes

Co-authored-by: Ralf Jung <post@ralfj.de>

workingjubilee committed:
Describe write_bytes and copy sans C

These intrinsics have a divergent meaning from C's own,
so stop trying to tell people that they are the same.

workingjubilee committed:
Remove scattered refs to C fn

Most docs are not actually referring to the symbol memcpy,
so do not refer to that symbol.

workingjubilee commented:
Note that I am leaving the aliases in place quite on-purpose: if someone is determined to search for that name, finding these functions and reading them is fine by me.

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -2603,6 +2598,8 @@ pub(crate) fn is_nonoverlapping<T>(src: *const T, dst: *const T, count: usize) -
 /// Note that even if the effectively copied size (`count * size_of::<T>()`) is
 /// `0`, the pointers must be non-null and properly aligned.
+/// In this case, the pointers may share the same address, as
+/// a region of 0 bytes overlaps with a region of 0 bytes for 0 bytes.
```
```suggestion
/// a region of 0 bytes does not overlap with a region of 0 bytes.
```
let's not try to be clever here...

RalfJung commented on library/core/src/intrinsics.rs:
```diff
@@ -2684,18 +2681,15 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
-/// `copy` is semantically equivalent to C's [`memmove`], but with the argument
-/// order swapped. Copying takes place as if the bytes were copied from `src`
-/// to a temporary array and then copied from the array to `dst`.
+/// `copy` may temporarily use additional space to copy bytes from `src` into an array, before
```
Doesn't it just copy forwards/backwards depending on which region is "lower"? I don't think you ever need a temporary array?

workingjubilee replied:
arguably a register is in fact a "temporary array", but you're right, I think.

RalfJung commented:
This reference to memset seems to still be around?

https://github.com/rust-lang/rust/blob/beb0e731de1c13ee0ee341965718b6350a48b34d/library/core/src/ptr/mut_ptr.rs#L1452

workingjubilee commented:
!@#$%^&*

workingjubilee committed:
Swat one last memset

workingjubilee committed:
Avoid winking description of non-overlap

Co-authored-by: Ralf Jung <post@ralfj.de>

RalfJung commented:
LGTM apart from that last open thread, feel free to resolve either way.

But let's also get more eyes on this since we are removing the C function equivalents entirely:
r? libs-api

workingjubilee commented:
Note that I am happy to readd documentation like what I removed if insisted-upon, but instead explaining more clearly that it is a [**false** friend](https://en.wikipedia.org/wiki/False_friend). The element vs. byte confusion, in particular, is something I will be landing a fix for in another Rust-based software repo. It is *really bad* out there!

the8472 commented:
The comment is there to some extent to tell people that if they're coming from C-land and would normally use memcpy/memmove/memset then these are the rusty replacements.
If the comment has to say "like memcpy, but different", that's fine.

RalfJung commented:
(Based on some discussion I had off-thread with @workingjubilee)

The thing is, there are many languages people come from, not just C. So should a comparison with C really be but front and center, before even our own list of safety requirements?

the8472 commented:
We do give C a bit of a special treatment here and there, since it's mentioned in the `#[alias]` attribute which directs people to the function if they're searching for that. It could probably be buried further down. After they have been directed there it would be good to to explicitly explain how it differs from memcpy.

workingjubilee commented:
Alright, I think that's reasonable.
@rustbot author

joshtriplett commented:
r? libs

Dylan-DPC commented:
@workingjubilee any updates on this? thanks

workingjubilee commented:
Somewhat hilariously, this PR has become much more incorrect, as it is almost exactly memcpy now that WG14 has declared an intent to validate `memcpy(NULL, NULL, 0)` in C2y.

Dylan-DPC commented:
Thanks :D. Closing this pr then :)
